### PR TITLE
Support fancy Option type constructors

### PR DESCRIPTION
**Pull request by pakrym**
Make the options constructor selection more resilient to ctor shape.

### CODE REVIEW

File: sdk/core/Microsoft.Extensions.Azure/tests/TestClientOptionsMultipleParameters.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        public ServiceVersion Version { get; }
+
+        public enum ServiceVersion
+        {
+            A,
+            B,
+            C,
+            D
+        }
+        public TestClientOptionsMultipleParameters(string otherParameter = "some default value", ServiceVersion version = ServiceVersion.D)
```
Comment (tg-msft) [inline sdk/core/Microsoft.Extensions.Azure/tests/TestClientOptionsMultipleParameters.cs]: Nit - ServiceVersion always has to be the first parameter per the guidelines

File: sdk/core/Microsoft.Extensions.Azure/src/Internal/ClientOptionsFactory.cs [hunk 2]
```diff
@@ -36,27 +36,53 @@ private TOptions CreateOptions(string name)
             }
 
             ConstructorInfo parameterlessConstructor = null;
-            ParameterInfo versionConstructor = null;
+            int versionParameterIndex = 0;
```
Comment (tg-msft) [inline sdk/core/Microsoft.Extensions.Azure/src/Internal/ClientOptionsFactory.cs]: Do you need to support this?  https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-versioning-select-api-version says the version always has to come first.

File: sdk/core/Microsoft.Extensions.Azure/src/Internal/ClientOptionsFactory.cs [hunk 3]
```diff
@@ -67,7 +93,17 @@ private TOptions CreateOptions(string name)
                 return Activator.CreateInstance<TOptions>();
             }
 
-            return (TOptions)Activator.CreateInstance(typeof(TOptions), versionConstructor.DefaultValue);
+            return (TOptions)Activator.CreateInstance(typeof(TOptions), constructorArguments);
+        }
+
+        private static bool IsServiceVersionParameter(ParameterInfo parameter)
+        {
+            if (parameter.ParameterType.Name == "ServiceVersion")
```
Comment (tg-msft) [inline sdk/core/Microsoft.Extensions.Azure/src/Internal/ClientOptionsFactory.cs]: Nit - maybe pull this into a constant?

**Review by tg-msft — APPROVED** [review]

File: sdk/core/Microsoft.Extensions.Azure/tests/TestClientOptionsMultipleParameters.cs (hunk 1 above)
Comment (pakrym) [reply sdk/core/Microsoft.Extensions.Azure/tests/TestClientOptionsMultipleParameters.cs]: People don't follow the guideline :) I'd rather be safe.

**Pull request merged**